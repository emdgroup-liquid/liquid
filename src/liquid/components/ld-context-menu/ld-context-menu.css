:host {
  display: block;
}

.ld-context-menu {
  list-style: none;
  border-radius: var(--ld-br-l);
  box-shadow: var(--ld-shadow-stacked);
  margin: 0;
  padding: var(--ld-sp-12);
}
