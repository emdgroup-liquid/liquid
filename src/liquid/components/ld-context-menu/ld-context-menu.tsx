import { Component, Element, h, Prop, State } from '@stencil/core'
import { isInnerFocusable } from '../../../liquid/utils/focus'

const isHtmlElement = (element?: Element): element is HTMLElement =>
  'focus' in element

@Component({
  tag: 'ld-context-menu',
  styleUrl: 'ld-context-menu.shadow.css',
  shadow: true,
})
export class LdContextMenu {
  @Element() el: HTMLLdContextMenuElement
  private menuRef: HTMLLdMenuElement
  private tooltipRef: HTMLLdTooltipElement
  private triggerSlotRef: HTMLSlotElement

  /** Position of the context menu relative to the trigger element. */
  @Prop() position?: HTMLLdTooltipElement['position'] = 'bottom left'

  /** Size of the context menu. */
  @Prop() size?: 'sm' | 'lg'

  @State() initialized = false

  private resetFocus = async () => {
    const [trigger] = this.triggerSlotRef.assignedElements()

    if (isInnerFocusable(trigger)) {
      await trigger.focusInner()
      return
    }

    if (isHtmlElement(trigger)) {
      trigger.focus()
    }
  }

  private handleKeyDown = async (event: KeyboardEvent) => {
    switch (event.key) {
      case 'Escape':
        event.preventDefault()
        await this.tooltipRef.hideTooltip()
    }
  }

  private handleMenuOpen = async () => {
    const tooltip = await this.tooltipRef.getTooltip()
    const menuInTooltip = tooltip.querySelector('ld-menu')

    if (!this.initialized) {
      menuInTooltip.addEventListener('keydown', this.handleKeyDown)
      this.initialized = true
    }
    const firstMenuItem = await menuInTooltip.getFirstMenuItem()

    if (!firstMenuItem) {
      return
    }

    await firstMenuItem.focusInner()
  }

  componentDidLoad() {
    const style = this.el.getAttribute('style')

    if (style) {
      this.menuRef.setAttribute('style', style)
      this.el.removeAttribute('style')
    }
  }

  render() {
    return (
      <ld-tooltip
        onLdtooltipclose={this.resetFocus}
        onLdtooltipopen={this.handleMenuOpen}
        ref={(element: HTMLLdTooltipElement) => (this.tooltipRef = element)}
        position={this.position}
        preventScreenreader
        tag="span"
        triggerType="click"
        unstyled
      >
        <slot
          name="trigger"
          ref={(element: HTMLSlotElement) => (this.triggerSlotRef = element)}
          slot="trigger"
        />
        <ld-menu ref={(el) => (this.menuRef = el)} size={this.size}>
          <slot />
        </ld-menu>
      </ld-tooltip>
    )
  }
}
