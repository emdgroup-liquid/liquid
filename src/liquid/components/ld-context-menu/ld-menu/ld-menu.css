:host {
  display: block;
}

.ld-menu {
  background: var(--ld-col-wht);
  border-radius: var(--ld-br-l);
  box-shadow: var(--ld-shadow-stacked);
  box-sizing: border-box;
  list-style: none;
  margin: 0;
  padding: var(--ld-sp-12);
}
