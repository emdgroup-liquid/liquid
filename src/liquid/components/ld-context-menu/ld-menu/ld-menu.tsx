import {
  Component,
  Element,
  h,
  Host,
  Method,
  Prop,
  State,
  Watch,
} from '@stencil/core'
import { isElement, isMenuItem, isSlot } from '../../../utils/type-checking'

const getMenuItemOrNestedMenuItems = (node: Node) => {
  if (!isElement(node)) {
    return []
  }

  if (isMenuItem(node)) {
    return [node]
  }

  const items: HTMLLdMenuitemElement[] = []

  if (isSlot(node)) {
    node
      .assignedNodes()
      .forEach((node) => items.push(...getMenuItemOrNestedMenuItems(node)))
    return items
  }

  node.childNodes.forEach((node) =>
    items.push(...getMenuItemOrNestedMenuItems(node))
  )

  return items
}

const isPrintableCharacter = (key: string) =>
  key.length === 1 && key.match(/\S/)

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part list - `ul` element wrapping the default slot
 */
@Component({
  tag: 'ld-menu',
  styleUrl: 'ld-menu.css',
  shadow: true,
})
export class LdMenu {
  @Element() el: HTMLLdMenuElement

  /** Size of the context menu. */
  @Prop() size?: 'sm' | 'lg'

  @State() initialized = false

  private initMenuItems = (element: Node, initial = false) => {
    if (!isElement(element)) {
      return
    }

    if (isMenuItem(element)) {
      element.size = this.size

      if (!initial) {
        return
      }

      element.ldTabindex = this.initialized ? -1 : 0

      if (!this.initialized) {
        this.initialized = true
      }

      return
    }

    if (isSlot(element)) {
      element.assignedNodes().forEach((node) => this.initMenuItems(node))
      return
    }

    element.childNodes.forEach((node) => this.initMenuItems(node))
  }

  /** Get the first menu item inside this menu. */
  @Method()
  async getFirstMenuItem(): Promise<HTMLLdMenuitemElement> {
    return this.getAllMenuItems()[0]
  }

  private getAllMenuItems = () => {
    const items: HTMLLdMenuitemElement[] = []

    this.el
      .querySelectorAll('slot, ld-menuitem')
      .forEach((node) => items.push(...getMenuItemOrNestedMenuItems(node)))

    return items
  }

  private focusFirst = (target: HTMLLdMenuitemElement) => {
    const allMenuItems = this.getAllMenuItems()
    const [first] = allMenuItems

    target.ldTabindex = -1
    first.ldTabindex = 0
    first.focusInner()
  }

  private focusLast = (target: HTMLLdMenuitemElement) => {
    const allMenuItems = this.getAllMenuItems()
    const last = allMenuItems[allMenuItems.length - 1]

    target.ldTabindex = -1
    last.ldTabindex = 0
    last.focusInner()
  }

  private focusNext = (target: HTMLLdMenuitemElement) => {
    const allMenuItems = this.getAllMenuItems()
    const index = allMenuItems.indexOf(target)
    const next =
      allMenuItems.length > index + 1
        ? allMenuItems[index + 1]
        : allMenuItems[0]

    target.ldTabindex = -1
    next.ldTabindex = 0
    next.focusInner()
  }

  private focusPrev = (target: HTMLLdMenuitemElement) => {
    const allMenuItems = this.getAllMenuItems()
    const index = allMenuItems.indexOf(target)
    const prev =
      index === 0
        ? allMenuItems[allMenuItems.length - 1]
        : allMenuItems[index - 1]

    target.ldTabindex = -1
    prev.ldTabindex = 0
    prev.focusInner()
  }

  private focusByFirstCharacter = (event: KeyboardEvent) => {
    if (!isPrintableCharacter(event.key)) {
      return
    }

    const target = event.target as HTMLLdMenuitemElement
    const allMenuItems = this.getAllMenuItems()
    const currentIndex = allMenuItems.indexOf(target)
    const itemMatchesKey = (item: HTMLLdMenuitemElement) =>
      item.textContent.trim()[0].toLocaleLowerCase() ===
      event.key.toLocaleLowerCase()
    let matchingItem = allMenuItems.find((item, index) => {
      if (index <= currentIndex) {
        return
      }

      return itemMatchesKey(item)
    })

    if (!matchingItem) {
      matchingItem = allMenuItems.find(itemMatchesKey)
    }

    if (!matchingItem || matchingItem === target) {
      return
    }

    target.ldTabindex = -1
    matchingItem.ldTabindex = 0
    matchingItem.focusInner()
  }

  private handleKeyDown = (event: KeyboardEvent) => {
    const target = event.target as HTMLLdMenuitemElement

    switch (event.key) {
      case 'ArrowUp':
        event.preventDefault()
        this.focusPrev(target)
        break
      case 'ArrowDown':
        event.preventDefault()
        this.focusNext(target)
        break
      case 'Home':
        event.preventDefault()
        this.focusFirst(target)
        break
      case 'End':
        event.preventDefault()
        this.focusLast(target)
        break
      default:
        this.focusByFirstCharacter(event)
    }
  }

  @Watch('size')
  handleSizeChange() {
    this.updateMenuItems()
  }

  private updateMenuItems = (initial = false) => {
    this.el
      .querySelectorAll('slot, ld-menuitem')
      .forEach((element) => this.initMenuItems(element, initial))
  }

  componentWillLoad() {
    this.updateMenuItems(true)
  }

  render() {
    return (
      <Host onKeyDown={this.handleKeyDown}>
        <ul class="ld-menu" part="list" role="menu">
          <slot />
        </ul>
      </Host>
    )
  }
}
