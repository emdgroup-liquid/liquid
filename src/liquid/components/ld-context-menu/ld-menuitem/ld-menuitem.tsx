import {
  Component,
  Element,
  Host,
  h,
  Prop,
  State,
  Method,
  Event,
  EventEmitter,
} from "@stencil/core";
import { cloneAttributes } from "../../../utils/cloneAttributes";

type Mode = "highlight" | "danger" | "neutral";

const modeMap = new Map<Mode, string>([
  ["danger", "danger-ghost"],
  ["highlight", "ghost"],
  ["neutral", "neutral-ghost"],
]);

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part listitem - `li` element wrapping the `ld-button` element
 * @part button - `ld-button` element wrapping the default slot
 */
@Component({
  tag: "ld-menuitem",
  styleUrl: "ld-menuitem.css",
  shadow: true,
})
export class LdMenuitem implements InnerFocusable {
  @Element() el: HTMLLdMenuitemElement;
  private attributesObserver: MutationObserver;
  private buttonRef?: HTMLLdButtonElement;

  /** Disabled state of the menu item. */
  @Prop() disabled?: boolean;

  /**
   * Transforms the menu item to an anchor element.
   * See [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-href)
   * for more information on the `href` attribute.
   */
  @Prop() href?: HTMLLdButtonElement["href"];

  /** Tab index of the menu item. */
  @Prop() ldTabindex?: number;

  /** Prevent closing of the context menu on click. */
  @Prop() preventClose?: boolean;

  /** Display mode. */
  @Prop() mode?: Mode = "neutral";

  /**
   * Size of the menu item.
   * @internal
   */
  @Prop() size?: "sm" | "lg";

  /**
   * The `target` attributed can be used in conjunction with the `href` attribute.
   * See [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target)
   * for more information on the `target` attribute.
   */
  @Prop() target?: HTMLLdButtonElement["target"];

  @State() clonedAttributes: Record<string, string>;

  /** Sets focus on the anchor or button */
  @Method()
  async focusInner() {
    this.buttonRef?.focusInner();
  }

  /**
   * @internal
   * Emitted on menu item click if preventClose prop is not truethy.
   */
  @Event() ldclosetooltip: EventEmitter;

  private handleClick = (ev: MouseEvent) => {
    if (this.preventClose) return;
    this.ldclosetooltip.emit(ev);
  };

  componentWillLoad() {
    this.attributesObserver = cloneAttributes.call(this, [
      "ld-tabindex",
      "mode",
      "size",
    ]);
  }

  disconnectedCallback() {
    /* istanbul ignore if */
    if (this.attributesObserver) this.attributesObserver.disconnect();
  }

  render() {
    return (
      <Host>
        <li class="ld-menuitem" part="listitem" role="menuitem">
          <ld-button
            {...this.clonedAttributes}
            class="ld-menuitem__button"
            disabled={this.disabled}
            href={this.href}
            iconOnly={false}
            justifyContent="start"
            ldTabindex={this.ldTabindex}
            mode={modeMap.get(this.mode)}
            onClick={this.handleClick}
            part="focusable button"
            ref={(element) => (this.buttonRef = element)}
            size={this.size}
            target={this.target}
            type="button"
          >
            <slot></slot>
          </ld-button>
        </li>
      </Host>
    );
  }
}
