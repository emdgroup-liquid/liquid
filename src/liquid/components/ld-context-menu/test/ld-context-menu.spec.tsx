import { h } from '@stencil/core'
import { AnyHTMLElement } from '@stencil/core/internal'
import { MockHTMLElement } from '@stencil/core/mock-doc'
import { newSpecPage, SpecPage } from '@stencil/core/testing'
import { LdContextMenu } from '../ld-context-menu'
import { LdMenuitem } from '../ld-menuitem/ld-menuitem'
import '../../../utils/mutationObserver'
import { LdTooltip } from '../../ld-tooltip/ld-tooltip'
import { LdMenu } from '../ld-menu/ld-menu'
import { LdButton } from '../../ld-button/ld-button'

const prepareAndGetMenuInTooltip = async (
  page: SpecPage,
  triggerSlotElements: Element[],
  tooltipContentNodes: Node[]
) => {
  const tooltip = page.root.shadowRoot.querySelector('ld-tooltip')
  const tooltipDefaultSlot = tooltip.shadowRoot.querySelector<AnyHTMLElement>(
    '.ld-tooltip__content slot'
  )
  const tooltipTriggerSlot = tooltip.shadowRoot.querySelector<AnyHTMLElement>(
    '.ld-tooltip__trigger slot'
  )
  const triggerSlot = page.root.shadowRoot.querySelector<AnyHTMLElement>('slot')
  const defaultSlot =
    page.root.shadowRoot.querySelector<AnyHTMLElement>('ld-menu slot')
  const menuItems = Array.from(page.root.querySelectorAll('ld-menuitem'))
  const triggerLdButton = page.root.querySelector('ld-button')
  const triggerButton = page.root.querySelector('button')

  triggerSlot.assignedElements = jest.fn().mockReturnValue(triggerSlotElements)
  tooltipTriggerSlot.assignedElements = jest.fn().mockReturnValue([triggerSlot])
  tooltipDefaultSlot.assignedNodes = jest
    .fn()
    .mockReturnValue(tooltipContentNodes)
  defaultSlot.assignedNodes = jest.fn().mockReturnValue(menuItems)

  jest.advanceTimersByTime(0)
  await page.waitForChanges()

  const menuInTooltip = tooltip.shadowRoot.querySelector(
    'ld-tooltip-popper ld-menu'
  )

  const menuItemsInTooltip = Array.from(
    menuInTooltip.querySelectorAll<HTMLLdMenuitemElement>('ld-menuitem')
  )

  const innerFocusableElements = [...menuItems, ...menuItemsInTooltip]

  if (triggerLdButton) {
    innerFocusableElements.push(triggerLdButton)
  }

  if (triggerButton) {
    triggerButton.focus = jest.fn()
  }

  innerFocusableElements.forEach((element) =>
    Object.defineProperty(element, 'focusInner', {
      value: jest.fn(),
      configurable: true,
      writable: true,
    })
  )

  Object.defineProperty(tooltip, 'hideTooltip', {
    value: jest.fn(),
    configurable: true,
    writable: true,
  })

  return menuInTooltip
}

describe('ld-context-menu', () => {
  let assignedElements: Element[]
  let assignedNodes: Node[]

  beforeAll(() => {
    Object.defineProperty(MockHTMLElement.prototype, 'assignedElements', {
      value: () => assignedElements ?? [],
      configurable: true,
      writable: true,
    })
    Object.defineProperty(MockHTMLElement.prototype, 'assignedNodes', {
      value: () => assignedNodes ?? [],
      configurable: true,
      writable: true,
    })
  })

  afterEach(() => {
    assignedElements = undefined
    assignedNodes = undefined
  })

  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem],
      template: () => (
        <ld-context-menu>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    expect(page.root).toMatchSnapshot()
  })

  it('forwards size prop to menu items in nested slots', async () => {
    // necessary to mock assignedNodes with a valid return value
    // TODO: these are actually no HTMLElements, maybe there is a better way?
    const menuItems = [
      <ld-menuitem>Menu item 1</ld-menuitem>,
      <ld-menuitem>Menu item 2</ld-menuitem>,
    ]
    const group = <ld-menuitem-group>{menuItems}</ld-menuitem-group>

    assignedNodes = [group]
    menuItems.forEach((menuItem) => {
      menuItem.tagName = 'LD-MENUITEM'
      menuItem.querySelectorAll = () => null
    })
    group.childNodes = menuItems
    group.querySelectorAll = () => null

    await newSpecPage({
      components: [LdContextMenu, LdMenu, LdMenuitem],
      template: () => (
        <ld-context-menu size="sm">
          <ld-button slot="trigger">Open</ld-button>
          {group}
        </ld-context-menu>
      ),
    })

    expect(menuItems[0].size).toBe('sm')
    expect(menuItems[1].size).toBe('sm')
  })

  it('forwards position', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem],
      template: () => (
        <ld-context-menu position="left middle">
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    expect(page.root).toMatchSnapshot()
  })

  it('clones the style attribute', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem],
      template: () => (
        <ld-context-menu style={{ width: '15rem' }}>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    expect(page.root).toMatchSnapshot()
  })

  it('focuses the first menu item when opened', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem, LdTooltip, LdMenu, LdButton],
      template: () => (
        <ld-context-menu>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    const tooltip = page.root.shadowRoot.querySelector('ld-tooltip')
    const tooltipTriggerButton =
      tooltip.shadowRoot.querySelector<HTMLLdButtonElement>(
        '.ld-tooltip__trigger'
      )
    const menu = page.root.shadowRoot.querySelector('ld-menu')
    const triggerButton = page.root.querySelector('ld-button')

    const menuInTooltip = await prepareAndGetMenuInTooltip(
      page,
      [triggerButton],
      [menu]
    )
    const firstMenuItemInTooltip = menuInTooltip.querySelector('ld-menuitem')

    // clicking the button inside the tooltip shadow DOM because event bubbling
    // doesn't work here, somehow.
    tooltipTriggerButton.click()
    await page.waitForChanges()

    expect(firstMenuItemInTooltip.focusInner).toHaveBeenCalled()
  })

  it('focuses the trigger ld-button when closed', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem, LdTooltip, LdMenu, LdButton],
      template: () => (
        <ld-context-menu>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    const tooltip = page.root.shadowRoot.querySelector('ld-tooltip')
    const tooltipTriggerButton =
      tooltip.shadowRoot.querySelector<HTMLLdButtonElement>(
        '.ld-tooltip__trigger'
      )
    const menu = page.root.shadowRoot.querySelector('ld-menu')
    const triggerButton = page.root.querySelector('ld-button')

    await prepareAndGetMenuInTooltip(page, [triggerButton], [menu])

    // clicking the button inside the tooltip shadow DOM because event bubbling
    // doesn't work here, somehow.
    tooltipTriggerButton.click()
    await page.waitForChanges()

    // trigger tooltip close
    const event = {
      type: 'touchend',
      isTrusted: true,
      composedPath: () => [page.body],
    }
    page.body.dispatchEvent(event as unknown as Event)

    expect(triggerButton.focusInner).toHaveBeenCalled()
  })

  it('focuses the trigger button when closed', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem, LdTooltip, LdMenu, LdButton],
      template: () => (
        <ld-context-menu>
          <button slot="trigger">Open</button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    const tooltip = page.root.shadowRoot.querySelector('ld-tooltip')
    const tooltipTriggerButton =
      tooltip.shadowRoot.querySelector<HTMLLdButtonElement>(
        '.ld-tooltip__trigger'
      )
    const menu = page.root.shadowRoot.querySelector('ld-menu')
    const triggerButton = page.root.querySelector('button')

    await prepareAndGetMenuInTooltip(page, [triggerButton], [menu])

    // clicking the button inside the tooltip shadow DOM because event bubbling
    // doesn't work here, somehow.
    tooltipTriggerButton.click()
    await page.waitForChanges()

    // trigger tooltip close
    const event = {
      type: 'touchend',
      isTrusted: true,
      composedPath: () => [page.body],
    }
    page.body.dispatchEvent(event as unknown as Event)

    expect(triggerButton.focus).toHaveBeenCalled()
  })

  it('closes context menu on Escape key', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem, LdTooltip, LdMenu, LdButton],
      template: () => (
        <ld-context-menu>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    const tooltip = page.root.shadowRoot.querySelector('ld-tooltip')
    const tooltipTriggerButton =
      tooltip.shadowRoot.querySelector<HTMLLdButtonElement>(
        '.ld-tooltip__trigger'
      )
    const menu = page.root.shadowRoot.querySelector('ld-menu')
    const triggerButton = page.root.querySelector('ld-button')

    const menuInTooltip = await prepareAndGetMenuInTooltip(
      page,
      [triggerButton],
      [menu]
    )
    const firstMenuItemInTooltip = menuInTooltip.querySelector('ld-menuitem')

    // clicking the button inside the tooltip shadow DOM because event bubbling
    // doesn't work here, somehow.
    tooltipTriggerButton.click()
    await page.waitForChanges()

    firstMenuItemInTooltip.dispatchEvent(
      new KeyboardEvent('keydown', { key: 'Escape', bubbles: true })
    )

    expect(tooltip.hideTooltip).toHaveBeenCalled()
  })

  it('prevents default on Tab key', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem, LdTooltip, LdMenu, LdButton],
      template: () => (
        <ld-context-menu>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    const tooltip = page.root.shadowRoot.querySelector('ld-tooltip')
    const tooltipTriggerButton =
      tooltip.shadowRoot.querySelector<HTMLLdButtonElement>(
        '.ld-tooltip__trigger'
      )
    const menu = page.root.shadowRoot.querySelector('ld-menu')
    const triggerButton = page.root.querySelector('ld-button')

    const menuInTooltip = await prepareAndGetMenuInTooltip(
      page,
      [triggerButton],
      [menu]
    )
    const firstMenuItemInTooltip = menuInTooltip.querySelector('ld-menuitem')
    const event = new KeyboardEvent('keydown', { key: 'Tab', bubbles: true })

    // clicking the button inside the tooltip shadow DOM because event bubbling
    // doesn't work here, somehow.
    tooltipTriggerButton.click()
    await page.waitForChanges()

    firstMenuItemInTooltip.dispatchEvent(event)

    expect(event.defaultPrevented).toBeTruthy()
  })

  it('allows to set inner focus on menu item', async () => {
    const page = await newSpecPage({
      components: [LdMenuitem],
      html: `<ld-menuitem />`,
    })
    const ldMenuitem = page.root
    const ldButton = ldMenuitem.shadowRoot.querySelector('ld-button')

    ldButton.focusInner = jest.fn()
    await ldMenuitem.focusInner()

    expect(ldButton.focusInner).toHaveBeenCalled()
  })

  it('allows to open and close menu via methods', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem, LdTooltip, LdMenu, LdButton],
      template: () => (
        <ld-context-menu>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })

    const contextMenu = page.root as HTMLLdContextMenuElement
    const tooltip = contextMenu.shadowRoot.querySelector('ld-tooltip')
    const menu = page.root.shadowRoot.querySelector('ld-menu')
    const triggerButton = page.root.querySelector('ld-button')
    const tooltipTrigger = tooltip.shadowRoot.querySelector(
      '.ld-tooltip__trigger'
    )
    const popper = tooltip.shadowRoot.querySelector('ld-tooltip-popper')

    await prepareAndGetMenuInTooltip(page, [triggerButton], [menu])

    expect(popper).toHaveAttribute('aria-hidden')

    await contextMenu.showContextMenu()
    await page.waitForChanges()

    expect(tooltipTrigger).toHaveClass('ld-tether-enabled')
    expect(popper).not.toHaveAttribute('aria-hidden')

    await contextMenu.hideContextMenu()
    await page.waitForChanges()

    expect(tooltip.hideTooltip).toHaveBeenCalled()
  })

  it('prevents closing menu', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem, LdTooltip, LdMenu, LdButton],
      template: () => (
        <ld-context-menu>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem prevent-close>Can not close</ld-menuitem>
          <ld-menuitem>Can close</ld-menuitem>
        </ld-context-menu>
      ),
    })

    const contextMenu = page.root as HTMLLdContextMenuElement
    const tooltip = page.root.shadowRoot.querySelector('ld-tooltip')
    const menu = page.root.shadowRoot.querySelector('ld-menu')
    const triggerButton = page.root.querySelector('ld-button')
    const popper = tooltip.shadowRoot.querySelector('ld-tooltip-popper')

    const menuInTooltip = await prepareAndGetMenuInTooltip(
      page,
      [triggerButton],
      [menu]
    )
    const menuItemsInTooltip = Array.from(
      menuInTooltip.querySelectorAll('ld-menuitem')
    )

    await contextMenu.showContextMenu()
    await page.waitForChanges()

    expect(popper).not.toHaveAttribute('aria-hidden')

    menuItemsInTooltip[0].click()
    await page.waitForChanges()

    expect(tooltip.hideTooltip).not.toHaveBeenCalled()

    menuItemsInTooltip[1].click()
    await page.waitForChanges()

    expect(tooltip.hideTooltip).not.toHaveBeenCalled()
  })

  it('opens on right-click', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem, LdTooltip, LdMenu, LdButton],
      template: () => (
        <ld-context-menu rightClick>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    const tooltip = page.root.shadowRoot.querySelector('ld-tooltip')
    const menu = page.root.shadowRoot.querySelector('ld-menu')
    const triggerButton = page.root.querySelector('ld-button')

    const menuInTooltip = await prepareAndGetMenuInTooltip(
      page,
      [triggerButton],
      [menu]
    )
    const firstMenuItemInTooltip = menuInTooltip.querySelector('ld-menuitem')

    tooltip.handleContextMenu(new Event('contextMenu'))
    await page.waitForChanges()

    expect(firstMenuItemInTooltip.focusInner).toHaveBeenCalled()
  })

  it('does not open on right-click if right-click is not enabled', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem, LdTooltip, LdMenu, LdButton],
      template: () => (
        <ld-context-menu>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    const tooltip = page.root.shadowRoot.querySelector('ld-tooltip')
    const menu = page.root.shadowRoot.querySelector('ld-menu')
    const triggerButton = page.root.querySelector('ld-button')

    const menuInTooltip = await prepareAndGetMenuInTooltip(
      page,
      [triggerButton],
      [menu]
    )
    const firstMenuItemInTooltip = menuInTooltip.querySelector('ld-menuitem')

    tooltip.handleContextMenu(new Event('contextMenu'))
    await page.waitForChanges()

    expect(firstMenuItemInTooltip.focusInner).not.toHaveBeenCalled()
  })

  it('closes on right-click of other context menu', async () => {
    const page = await newSpecPage({
      components: [LdContextMenu, LdMenuitem, LdTooltip, LdMenu, LdButton],
      template: () => (
        <ld-context-menu rightClick>
          <ld-button slot="trigger">Open</ld-button>
          <ld-menuitem>Menu item</ld-menuitem>
        </ld-context-menu>
      ),
    })
    const tooltip = page.root.shadowRoot.querySelector('ld-tooltip')
    const menu = page.root.shadowRoot.querySelector('ld-menu')
    const triggerButton = page.root.querySelector('ld-button')

    const menuInTooltip = await prepareAndGetMenuInTooltip(
      page,
      [triggerButton],
      [menu]
    )
    const firstMenuItemInTooltip = menuInTooltip.querySelector('ld-menuitem')

    tooltip.handleContextMenu(new Event('contextMenu'))
    await page.waitForChanges()

    expect(firstMenuItemInTooltip.focusInner).toHaveBeenCalled()

    const closeSpy = jest.fn()
    tooltip.addEventListener('ldtooltipclose', closeSpy)

    page.body.dispatchEvent(new CustomEvent('ldtooltipopen', { bubbles: true }))
    await page.waitForChanges()

    expect(closeSpy).toHaveBeenCalled()
  })

  it('does not throw when trying to set inner focus before hydration', async () => {
    const component = new LdMenuitem()
    await component.focusInner()
  })

  it('does not throw when disconnecting before hydration', () => {
    const component = new LdMenuitem()
    component.disconnectedCallback()
  })
})
