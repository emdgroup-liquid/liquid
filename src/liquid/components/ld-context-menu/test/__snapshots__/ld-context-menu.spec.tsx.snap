// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-context-menu clones the style attribute 1`] = `
<ld-context-menu>
  <mock:shadow-root>
    <ld-tooltip position="bottom left" triggertype="click" unstyled="">
      <slot name="trigger" slot="trigger"></slot>
      <ld-menu style="width: 15rem;">
        <slot></slot>
      </ld-menu>
    </ld-tooltip>
  </mock:shadow-root>
  <ld-button slot="trigger">
    Open
  </ld-button>
  <ld-menuitem>
    <mock:shadow-root>
      <li class="ld-menuitem" role="menuitem">
        <ld-button class="ld-menuitem__button" justifycontent="start" mode="neutral-ghost" type="button">
          <slot></slot>
        </ld-button>
      </li>
    </mock:shadow-root>
    Menu item
  </ld-menuitem>
</ld-context-menu>
`;

exports[`ld-context-menu forwards position 1`] = `
<ld-context-menu>
  <mock:shadow-root>
    <ld-tooltip position="left middle" triggertype="click" unstyled="">
      <slot name="trigger" slot="trigger"></slot>
      <ld-menu>
        <slot></slot>
      </ld-menu>
    </ld-tooltip>
  </mock:shadow-root>
  <ld-button slot="trigger">
    Open
  </ld-button>
  <ld-menuitem>
    <mock:shadow-root>
      <li class="ld-menuitem" role="menuitem">
        <ld-button class="ld-menuitem__button" justifycontent="start" mode="neutral-ghost" type="button">
          <slot></slot>
        </ld-button>
      </li>
    </mock:shadow-root>
    Menu item
  </ld-menuitem>
</ld-context-menu>
`;

exports[`ld-context-menu renders 1`] = `
<ld-context-menu>
  <mock:shadow-root>
    <ld-tooltip position="bottom left" triggertype="click" unstyled="">
      <slot name="trigger" slot="trigger"></slot>
      <ld-menu>
        <slot></slot>
      </ld-menu>
    </ld-tooltip>
  </mock:shadow-root>
  <ld-button slot="trigger">
    Open
  </ld-button>
  <ld-menuitem>
    <mock:shadow-root>
      <li class="ld-menuitem" role="menuitem">
        <ld-button class="ld-menuitem__button" justifycontent="start" mode="neutral-ghost" type="button">
          <slot></slot>
        </ld-button>
      </li>
    </mock:shadow-root>
    Menu item
  </ld-menuitem>
</ld-context-menu>
`;
